### PR TITLE
Add filter associations. Fixes #45

### LINKED ISSUE #45: Add filters_association that allow specification of multiple parts within an aggregation
Reference: https://github.com/projecthydra-labs/hydra-pcdm/wiki/Proposed-Interfaces#defining-an-aggregation-with-multiple-filters

Example definition:
```ruby
aggregates :members, predicate: RDFVocabularies::PCDMTerms.hasMember,
           class_name: "ActiveFedora::Base"

filters_association :members, as: :child_objects, condition: :pcdm_object?
filters_association :members, as: :child_collections, condition: :pcdm_collection?
```

By defining a filters_association, you get the following methods defined that provide getters/setters with the method name in as: and validation using the method specified in condition:.

```ruby
child_objects=    # validate with pcdm_object?
child_objects+=   # validate with pcdm_object?
child_objects<<   # validate with pcdm_object?
child_objects     # filter to return only objects

child_collections=    # validate with pcdm_collection?
child_collections+=   # validate with pcdm_collection?
child_collections<<   # validate with pcdm_collection?
child_collections     # filter to return only collections

members=    # validate with pcdm_object and pcdm_collection?
members+=   # validate with pcdm_object and pcdm_collection?
members<<   # validate with pcdm_object and pcdm_collection?
members     # return everything in the aggregation
```

### PR DESCRIPTION
**Pull request by jcoyne**

### CODE REVIEW

File: lib/active_fedora/aggregation/base_extension.rb [hunk 1]
```diff
@@ -29,11 +31,28 @@ def aggregates(name, options={})
         Builder.build(self, name, options)
       end
 
+      ##
+      # Create a association filter on the class
```
Comment (mjgiarlo) [inline lib/active_fedora/aggregation/base_extension.rb]: `s/a association/an association/`

File: lib/active_fedora/aggregation/base_extension.rb [hunk 2]
```diff
@@ -29,11 +31,28 @@ def aggregates(name, options={})
...
+      # @example
+      #   class Image < ActiveFedora::Base
+      #     aggregates :generic_files
+      #     filters_association :generic_files, as: :large_files, condition: :is_big?
```
Comment (mjgiarlo) [inline lib/active_fedora/aggregation/base_extension.rb]: This is super nitpicky but I don't like encouraging people to name methods with `is_` ;)

File: lib/active_fedora/filter/association.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+      def find_target?
+        true
+      end
+
+      def find_target
+        existing_matching_records
+      end
+
+      # We can't create an association scope on here until we can figure a way to
```
Comment (mjgiarlo) [inline lib/active_fedora/filter/association.rb]: Do we need, or already have, a ticket for this?

File: lib/active_fedora/filter/association.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        owner.association(options.fetch(:extending_from))
+      end
+
+      def validate_assertion(record)
+        record.send(options.fetch(:condition))
+      end
+
+      def validate_assertion!(record)
+        raise ArgumentError, "#{record.class} with ID: #{record.id} was expected to #{options.fetch(:condition)}, but it was false" unless validate_assertion(record)
+
```
Comment (mjgiarlo) [inline lib/active_fedora/filter/association.rb]: Blank line?

File: lib/active_fedora/filter/association.rb (hunk 3 above)
Comment (jcoyne) [inline lib/active_fedora/filter/association.rb]: I don't think we need one, it's just an explanation for why the method exists.

**Pull request merged**